### PR TITLE
Adding CustomPainterComponent

### PR DESCRIPTION
**Pull request by erickzanardo**
# Description

Adds `CustomPainterComponent`, a component which can help sharing custom rendering logic between Flutter widgets and a Flame game.

https://user-images.githubusercontent.com/835641/139759709-b3b252b9-607c-4cb2-919d-9a9fc008d14e.mov

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ x I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] My PR does not decrease the code coverage, or I have __a very special case__ and explained on the PR description why this PR decreases the coverage.
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `./scripts/format.sh` and the Flame analyzer (`./scripts/analyze.sh`) does not report any problems.
- [x] I read and followed the [Flame Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [x] No, this is *not* a breaking change.

## Related Issues

Solves: #1028 

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

File: doc/components.md [hunk 1]
```diff
@@ -492,6 +492,21 @@ Check the example app
 [nine_tile_box](https://github.com/flame-engine/flame/blob/main/examples/lib/stories/utils/nine_tile_box.dart)
 for details on how to use it.
 
+## CustomPainterComponent
+
+A `CustomPainter` is a useful Flutter class used with the `CustomPaint` widget to render custom
```
Comment (spydon) [inline doc/components.md:497]:
```suggestion
A `CustomPainter` is a Flutter class used with the `CustomPaint` widget to render custom
```
We only add useful classes ;)

File: doc/components.md [hunk 2]
```diff
@@ -492,6 +492,21 @@ Check the example app
...
+shapes inside a Flutter application.
+
+Flame provides a component that can render a `CustomPainter` called `CustomPainterComponent`, it
+receives a custom painter and just render it on the game canvas.
```
Comment (spydon) [inline doc/components.md:501]:
```suggestion
receives a custom painter and just renders it on the game canvas.
```

File: packages/flame/lib/src/components/custom_painter_component.dart [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+/// This component is usefull to [CustomPainter]s between your Flutter UI and Flame game.
+class CustomPainterComponent extends PositionComponent {
+  /// The [CustomPainter] used to render this component
+  CustomPainter? painter;
+
+  CustomPainterComponent({
+    Vector2? position,
+    Vector2? size,
+    int? priority,
+    this.painter,
```
Comment (spydon) [inline packages/flame/lib/src/components/custom_painter_component.dart:19]: Should we not have this as `@required` and the global field as non-nullable?

Comment (spydon) [reply packages/flame/lib/src/components/custom_painter_component.dart:19]: I would have this as the first argument too.

**Review by spydon — APPROVED** [review]
Lgtm, some very minor comments

File: doc/components.md (hunk 2 above)
Comment (wolfenrain) [reply doc/components.md:501]: I would say that the word `just` can also be removed, it adds no extra information and breaks up the sentence unnecessarily imho.

File: doc/components.md [hunk 4]
```diff
@@ -492,6 +492,21 @@ Check the example app
...
+
+This can be useful for sharing custom rendering logic between your Flame game, and your Flutter
```
Comment (wolfenrain) [inline doc/components.md:503]:
```suggestion
This can be used for sharing custom rendering logic between your Flame game, and your Flutter
```

**Review by wolfenrain — APPROVED** [review]

File: doc/components.md (hunk 2 above)
Comment (spydon) [reply doc/components.md:501]: Ageed, edited the suggestion.

File: packages/flame/lib/src/components/custom_painter_component.dart (hunk 3 above)
Comment (erickzanardo) [reply packages/flame/lib/src/components/custom_painter_component.dart:19]: I am following the way we do on sprite and sprite animation component.

Keeping this as optional we allow people to both pass the painter on the constructor, or when extending this class, init it on the `onLoad` method. Which is what we have been doing latelly.

File: packages/flame/lib/src/components/custom_painter_component.dart [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+import '../extensions/vector2.dart';
+import 'position_component.dart';
+
+/// A [PositionComponent] that renders a [CustomPainter] at the designated
+/// position, scaled to have the designated size and rotated to the specified
+/// angle.
+///
+/// This component is usefull to [CustomPainter]s between your Flutter UI and Flame game.
+class CustomPainterComponent extends PositionComponent {
```
Comment (renancaraujo) [inline packages/flame/lib/src/components/custom_painter_component.dart:11]: We should highlight that should repaint is ignored.

**Review by renancaraujo — APPROVED** [review]

File: packages/flame/lib/src/components/custom_painter_component.dart (hunk 5 above)
Comment (erickzanardo) [reply packages/flame/lib/src/components/custom_painter_component.dart:14]: Done

**Pull request merged**